### PR TITLE
Implement Virtual Scroll to load 7k sites

### LINKED ISSUE #28843: Implement Virtual Scroll to load 6k sites
Parent Issue

https://github.com/dotCMS/core/issues/28494

### Current behavior

The actual field when there are more than 6k sites is so slow and the TreeSelect does not work well, the idea is to clone the TreeSelect and implement virtual scrolling to avoid this behavior.

### Video Support

https://github.com/dotCMS/core/assets/7611944/a8d1dadb-27cd-4a1d-a5ab-8b7aff655887

### Steps for IQA

1. Ask Nicolas Molina or Daniel Silva for a db with 7k sites.
2. Load this db into your docker db 

```
# Copy the db to your container
docker cp ./wth.sql postgres-dotcms:./
# Load this db
docker exec -it postgres-dotcms psql -U dotcms -f ./wth.sql <your_db>
```

3. Enable `DOT_FEATURE_FLAG_NEW_EDIT_PAGE` in your docker compose.
 
Screenshot 2024-06-17 at 12 10 45 PM

4. Create a new content type using the new editor content

Screenshot 2024-06-17 at 12 12 10 PM

5. Add a SiteOrFolder field

Screenshot 2024-06-17 at 12 08 03 PM

6. Test this field by creating new content based on this schema

Screenshot 2024-06-17 at 12 13 24 PM

### PR DESCRIPTION
**Pull request by nicobytes**
### Proposed Changes

- Improve tree component story using virtual scroll and strong typing
- Fix bug in DotSiteService with `perPage` param and take advantage to refactor unit tests with spectator
- Clone TreeSelect component locally because in v15 this component doesn't have the inputs to enable virtual scroll
- Refactor DotEditContentHostFolderFieldComponent using ngrx/signals to handle all state based on signals

### Checklist
- [x] Tests
- [x] Translations
- [x] Security Implications Contemplated (add notes if applicable)

This PR fixes: #28843

### CODE REVIEW

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+                            return item;
+                        });
+                        changes.tree = newTree;
+                    }
+
+                    patchState(store, changes);
+                })
+            )
+        ),
```
Comment (zJaaal) [inline core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts:100]: Can we granulate this in different functions? Looks really hard to read and maintain

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                exhaustMap((event: TreeNodeSelectItem) => {
+                    const { node } = event;
+                    const { hostname, path } = node.data;
+
+                    return dotEditContentService.getFoldersTreeNode(hostname, path).pipe(
+                        tap((children) => {
+                            node.leaf = true;
+                            node.icon = 'pi pi-folder-open';
+                            node.children = children;
+                            node.expanded = true;
```
Comment (zJaaal) [inline core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts:112]: can we use an object and do only one assign?

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    TreeNodeItem,
+    TreeNodeSelectItem
+} from '../../../models/dot-edit-content-host-folder-field.interface';
+import { DotEditContentService } from '../../../services/dot-edit-content.service';
+
+type HostFolderFiledState = {
+    nodeSelected: TreeNodeItem | null;
+    nodeExpaned: TreeNodeSelectItem['node'] | null;
+    tree: TreeNodeItem[];
+    status: 'idle' | 'pending' | 'fulfilled' | { error: string };
```
Comment (zJaaal) [inline core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts:21]: should we add another prop for the error? Looks ambiguous to have an error object inside a property that is mostly an string

**Review by zJaaal — DISMISSED** [review]

File: core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts [hunk 4]
```diff
@@ -17,21 +17,20 @@ export interface SiteParams {
     providedIn: 'root'
 })
 export class DotSiteService {
-    private BASE_SITE_URL = '/api/v1/site';
-    private params: SiteParams = {
+    #BASE_SITE_URL = '/api/v1/site';
```
Comment (oidacra) [inline core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts:20]: Can we use a constant for this? to share the constant in the service and test, if the endpoint changes, the test will fail.

File: core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts [hunk 5]
```diff
@@ -42,21 +41,21 @@ export class DotSiteService {
      * @memberof DotSiteService
      */
     getSites(filter = '*', perPage?: number): Observable<Site[]> {
```
Comment (oidacra) [inline core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts:43]: `perPage?: DEFAULT_PER_PAGE`  constant

File: core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts [hunk 6]
```diff
@@ -42,21 +41,21 @@ export class DotSiteService {
...
-        return this.http
-            .get<{ entity: Site[] }>(this.siteURL(filter, perPage))
+        return this.#http
```
Comment (oidacra) [inline core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts:44]: We usually use URLSearchParams for handle the params.
```
const searchParams = new URLSearchParams({
            filter,
            per_page: perPage || PER_PAGE,
            ...this.defaultParams
        });

return this.http.get<{ entity: Site[] }>(`
          ${BASE_SITE_URL}?${searchParams.toString()}
          `).pipe(pluck('entity'));
```

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts (hunk 3 above)
Comment (oidacra) [reply core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts:21]: the message of error is not a status. Maybe have `status` (with 'error') and `error: string` in the store. 
We have centralized this constant
```
// core-web/libs/dotcms-models/src/lib/shared-models.ts
export enum ComponentStatus {
    INIT = 'INIT',
    LOADING = 'LOADING',
    LOADED = 'LOADED',
    SAVING = 'SAVING',
    IDLE = 'IDLE'
}
```

We could add `ERROR` and use from it.

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            };
+        })
+    })),
+    withMethods((store, dotEditContentService = inject(DotEditContentService)) => ({
+        loadSites: rxMethod<string | null>(
+            pipe(
+                tap(() => patchState(store, { status: 'pending' })),
+                switchMap((path) => {
+                    return dotEditContentService
+                        .getSitesTreePath({ perPage: 7000, filter: '*' })
```
Comment (oidacra) [inline core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts:49]: this `7000` convert it to a constant

File: core-web/libs/edit-content/src/lib/services/dot-edit-content.service.ts [hunk 8]
```diff
@@ -74,8 +75,10 @@ export class DotEditContentService {
      * @return {*}  {Observable<TreeNodeItem[]>}
      * @memberof DotEditContentService
      */
-    getSitesTreePath(): Observable<TreeNodeItem[]> {
-        return this.#siteService.getSites().pipe(
+    getSitesTreePath(data = { filter: '*', perPage: 7000 }): Observable<TreeNodeItem[]> {
```
Comment (oidacra) [inline core-web/libs/edit-content/src/lib/services/dot-edit-content.service.ts:78]: use the constant (7000)

Comment (oidacra) [reply core-web/libs/edit-content/src/lib/services/dot-edit-content.service.ts:78]: What happend if the client has more than 7000?

File: core-web/libs/data-access/src/lib/dot-site/dot-site.service.spec.ts [hunk 9]
```diff
@@ -1,69 +1,54 @@
...
 import { DotSiteService, SiteParams } from './dot-site.service';
 
 describe('DotSiteService', () => {
+    let spectator: SpectatorHttp<DotSiteService>;
     let service: DotSiteService;
-    let httpTestingController: HttpTestingController;
+
+    const createHttp = createHttpFactory(DotSiteService);
 
     const BASE_SITE_URL = '/api/v1/site';
```
Comment (oidacra) [inline core-web/libs/data-access/src/lib/dot-site/dot-site.service.spec.ts:13]: Centralize this constant, and reuse it in the service and test.

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts (hunk 1 above)
Comment (nicobytes) [reply core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts:100]: @zJaaal Yes on the same page with you I think is a lot of functions in one file, but the `ngrx/signals` works only with closure functions with `store` in all context, check this example.

Screenshot 2024-06-14 at 12 53 33 PM

Here is an example to turn the store into a class, but the methods with rxMethod cannot be private methods instead should be treated like attrs.

Screenshot 2024-06-14 at 12 57 32 PM

I need to do more research on this and create a standard with the frontend team.

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts (hunk 2 above)
Comment (nicobytes) [reply core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts:112]: @zJaaal done!

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts (hunk 3 above)
Comment (nicobytes) [reply core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts:21]: @oidacra @zJaaal done

File: core-web/libs/data-access/src/lib/dot-site/dot-site.service.spec.ts (hunk 9 above)
Comment (nicobytes) [reply core-web/libs/data-access/src/lib/dot-site/dot-site.service.spec.ts:13]: @oidacra done!

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts (hunk 7 above)
Comment (nicobytes) [reply core-web/libs/edit-content/src/lib/fields/dot-edit-content-host-folder-field/store/host-folder-field.store.ts:49]: @oidacra done!

File: core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts (hunk 4 above)
Comment (nicobytes) [reply core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts:20]: @oidacra done!

File: core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts (hunk 6 above)
Comment (nicobytes) [reply core-web/libs/data-access/src/lib/dot-site/dot-site.service.ts:44]: @oidacra done!

**Review by oidacra — APPROVED** [review]

**Review by KevinDavilaDotCMS — APPROVED** [review]

**Pull request merged**